prow/jira/jira.go

```diff
@@ -251,16 +291,287 @@ func (jc *client) UpdateRemoteLink(id string, link *jira.RemoteLink) error {
+	var nameList []string
+	for _, transition := range transitions {
+		// JIRA shows all statuses as caps in the UI, but internally has different case; use EqualFold to ignore case
+		if strings.EqualFold(transition.Name, statusName) {
+			transitionID = transition.ID
+			break
+		}
+		nameList = append(nameList, transition.Name)
+	}
+	if transitionID == "" {
+		return fmt.Errorf("No transition status with name `%s` could be found. Please select from the following list: %v", statusName, nameList)
+	}
+	return jc.DoTransition(issueID, transitionID)
+}
+
+func (jc *client) UpdateStatus(issueID, statusName string) error {
+	return UpdateStatus(jc, issueID, statusName)
+}
+
+func (jc *client) GetTransitions(issueID string) ([]jira.Transition, error) {
+	transitions, resp, err := jc.upstream.Issue.GetTransitions(issueID)
+	if err != nil {
+		return nil, HandleJiraError(resp, err)
+	}
+	return transitions, nil
+}
+
+func (jc *client) DoTransition(issueID, transitionID string) error {
+	resp, err := jc.upstream.Issue.DoTransition(issueID, transitionID)
+	if err != nil {
+		return HandleJiraError(resp, err)
+	}
+	return nil
+}
+
+func (jc *client) UpdateIssue(issue *jira.Issue) (*jira.Issue, error) {
+	result, resp, err := jc.upstream.Issue.Update(issue)
+	if err != nil {
+		return nil, HandleJiraError(resp, err)
+	}
+	return result, nil
+}
+
+func (jc *client) AddComment(issueID string, comment *jira.Comment) (*jira.Comment, error) {
+	result, resp, err := jc.upstream.Issue.AddComment(issueID, comment)
+	if err != nil {
+		return nil, HandleJiraError(resp, err)
+	}
+	return result, nil
+}
+
+func (jc *client) CreateIssue(issue *jira.Issue) (*jira.Issue, error) {
+	result, resp, err := jc.upstream.Issue.Create(issue)
+	if err != nil {
+		return nil, HandleJiraError(resp, err)
+	}
+	return result, nil
+}
+
+func (jc *client) CreateIssueLink(link *jira.IssueLink) error {
+	resp, err := jc.upstream.Issue.AddLink(link)
+	if err != nil {
+		return HandleJiraError(resp, err)
+	}
+	return nil
+}
+
+// CloneIssue copies an issue struct, clears unsettable fields, creates a new
+// issue using the updated struct, and then links the new issue as a clone to
+// the original.
+func CloneIssue(jc Client, parent *jira.Issue) (*jira.Issue, error) {
+	// create deep copy of parent "Fields" field
+	data, err := json.Marshal(parent.Fields)
+	if err != nil {
+		return nil, err
+	}
+	childIssueFields := &jira.IssueFields{}
+	err = json.Unmarshal(data, childIssueFields)
+	if err != nil {
+		return nil, err
+	}
+	childIssue := &jira.Issue{
+		Fields: childIssueFields,
+	}
+	// update description
+	childIssue.Fields.Description = fmt.Sprintf("This is a clone of issue %s. The following is the description of the original issue: \n---\n%s", parent.Key, parent.Fields.Description)
+
+	// attempt to create the new issue
+	createdIssue, err := jc.CreateIssue(childIssue)
+	if err != nil {
+		// some fields cannot be set on creation; unset them
+		if JiraErrorStatusCode(err) == 400 {
```

Comment:
Nit: Easier to read by using early returns to avoid indentation, i.E:
```suggestion
		if JiraErrorStatusCode(err) != 400 {
		  return nil, err
		}
		
		// http 400 handling continues here and doesn't need nesting
```